### ISSUE TITLE
Time Zone Support

### ISSUE THREAD

**Issue by joeydong** [jakubroztocil/rrule#38]
Is there a way to get a `.between` for an RRULE that is in a time zone other than the browser's current time zone? could the library make use of [timezone-js](https://github.com/mde/timezone-js) to implement this functionality?

For example, if I'm in America/New_York but I have a recurrence that should be interpreted in Europe/Paris, the resulting dates should account for the daylight savings changes in the Europe/Paris time zone.

Edit: Digging deeper, would a change to

```
tzOffset: function(date) {
  return date.getTimezoneOffset() * 60 * 1000
}
```

fix this issue?

**Comment by visionscaper** [comment]
Very interested as well!

**Comment by aramk** [comment]
I've added #97 to add support for time zones. It needed some tinkering across the entire library but I've added tests and think it's a pretty decent solution. I'll be using it in production shortly.

**Comment by Yatufo** [comment]
I used the workaround using http://momentjs.com/timezone/
var options = RRule.parseString(RULE_STRING_WITH_DT_START)
options.dtstart = moment.tz(ORIGINAL_DATE, TIME_ZONE).toDate();
var rule = new RRule(options)
Cheers

**Comment by aramk** [comment]
@Yatufo I'm fairly certain that using toDate() will convert the date into the local time zone, which is all JavaScript's Date supports. This is fine for displaying the local date to users, but recurrences generated in this way will have issues due to time zone dependent `dtstart`.

**Comment by jamesdixon** [comment]
Any chance @aramk's solution (or another) gets merged? I'm running into the issue where my recurrences are changing when encountering Daylight Savings Time

**Comment by s-taylor** [comment]
For anyone who needs it, I was able to hack around the problem by applying an offset to rrules dates AFTER they are generated. We're using node.js and rrule defaults to using our server time which is in UTC but we want the occurences generated in Australia/Sydney time. This is how I solved it...
https://tonicdev.com/nizmox/576896ef64545613002b49b1

**Comment by mchapman** [comment]
My [fork](https://github.com/reallycare/rrule), which borrowed heavily from others, is in production.  Would really love a more widely adopted solution, but now I don't even know whether the best repo to do that is here or in the @arolson101 fork

**Comment by s-taylor** [comment]
As an experiment, I attempted to build my own recurrance rule library a while back leveraging moment-timezone, and it was fairly successful https://www.npmjs.com/package/time-walk. It's a fair bit different from rrule though, so may not work for everyone and probably needs more work to be production ready.

**Comment by jimmywarting** [comment]
vanilla `Intl.DateTimeFormat` is probably the best solution for converting dates between different timezones?

**Comment by davidgoli** [comment]
Would love an update on this issue as it's a fairly serious deviation from both the RRULE RFC spec and the Python implementation. @aramk 's fork seems to have a workable solution AFAICT, though it is a bit old. Any updates?

**Comment by davidgoli** [comment]
@jimmywarting `Intl.DateTimeFormat` is *not quite* widely supported, particularly on Android browser: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat#Browser_compatibility

**Comment by davidgoli** [comment]
@jakubroztocil any reason `tzinfos` and `ignoretz` are collected (and documented) but not used? Seems like line 627 is the spot for it:

```js
            // python:
            // datetime.time(hour, minute, second,
            // tzinfo=self._tzinfo))
            this.timeset.push(new dateutil.Time(hour, minute, second, millisecondModulo))
```

**Comment by davidgoli** [comment]
Also seems like implementing python's more recent tests might be a good first approach: https://github.com/dateutil/dateutil/blob/97dd3697df28bc50baaa2a098124a7428218f74e/dateutil/test/test_rrule.py#L4548

**Comment by davidgoli** [comment]
as it turns out, by not running in a UTC environment (eg with `TZ=UTC npm test`, existing tests fail. Fixing them should already get the implementation pretty far along.

**Comment by thefliik** [comment]
@davidgoli fyi, this library is not actively maintained at the moment

**Comment by davidgoli** [comment]
@thefliik that certainly seems to be the case! but there's no indication on the README that this is so. Are you a maintainer?

**Comment by davidgoli** [comment]
related: do you know of any alternatives that _are_ maintained that implement the RFC?

**Comment by thefliik** [comment]
@davidgoli  I'm not a maintainer. I believe github always adds a tag next to someone's name if they've contributed to a project (or is officially involved with it).

And I've looked for alternatives and I'm fairly confident in saying this is still the best opensource javascript based recurrence library.

**Comment by davidgoli** [comment]
@jakubroztocil @arolson101 care to weigh in? Is this library officially abandoned, or are there plans to take submissions from the community?

**Comment by thefliik** [comment]
This library is *not* officially abandoned, and I didn't mean to imply that it was. This library *is not actively maintained*. I imagine at some point in the future, PR's will be accepted again. I just wanted to chime in because it's quite likely you're specific issue will not be fixed in a timely manner. If that's a dealbreaker for you, you should look elsewhere. Of course, one of the maintainers is free to correct me if I'm wrong.

**Comment by davidgoli** [comment]
That's good to know, the question we're struggling with is not whether to adopt this library (our sunk costs are already too great for that), but rather whether to fork and patch or to try and get the fix merged. We'd rather not maintain our own fork if we can help it.

**Comment by thefliik** [comment]
Having watched this library for a while now. (Over a year? Maybe) I'm pretty confident in saying you should fork and patch. Definitely look at some of the forks others have already made. And, obviously, a maintainer may correct me, but just [taking a look at the commit history](https://github.com/jakubroztocil/rrule/graphs/commit-activity) can give you a sense of activity on the repo.

**Comment by davidgoli** [comment]
That's fair - combined with the fact that this issue is nearing 5 years old at this point. It's a bit surprising, I would think there would be bigger stakeholders willing to either implement their own rrule library or support one like this. I wonder if it's just the case that companies try to avoid building calendar-type apps? or are there more popular alternatives nowadays?

**Comment by davidgoli** [comment]
@thefliik Thanks for all the context. Granted, this `rrule` library _could_ work if we committed to only using it server-side since we could run it under `UTC` and all would be well. However, we're trying to use it client-side for performance reasons (a user's list of events is orders of magnitude smaller and more cacheable compared to the number of recurrences), so we're hoping a patch might be considered. It looks like there was activity here as recently as February, so while we're not expecting a quick turnaround, we're holding out hope there might be a possibility of a merge within weeks or maybe a few months.

**Comment by jgornick** [comment]
It looks like there are other forks that added TZ support too but just haven't contributed back. 

Is the issue that this project is not active and there isn't a place for this TZ work to land?

Does it make sense then to ask @jakubroztocil for others to become maintainers or to have an official fork?

**Comment by thefliik** [comment]
@jgornick you might want to read through the full conversation in issue #160. As I just mentioned, this library already has ~5 maintainers I believe (much more than most libraries I've used with this many stars). The issue is that adding TZ support is a large update and no one has the time to deal with it this year. Or last year. I also don't think there are any forks that have a complete TZ implementation (though feel free to prove me wrong). `@jakubroztocil` has made clear he doesn't have time for this library at the moment (which again, is completely fair). You can read #160 for some more context.

**Comment by jamesdixon** [comment]
@davidgoli we are using this library client side for the same reasons you mention. I can get you details but I believe we basically strip off the `Z` on everything and deal with it in local time. Let me know if you need details and I'll see what I can do

**Comment by jamesdixon** [comment]
@davidgoli are you using Ember?

**Comment by davidgoli** [comment]
@jamesdixon Current project is using React Native. For our purposes, we'd like to deal with everything in UTC "zoneless" time and simply present it in whatever timezone the user happens to be in currently. The issue we're seeing in particular is that because this library uses JS `Date`, which in turn has the runtime timezone baked in, we get daylight savings time applied around the local DST date boundary even though UTC doesn't have DST. So for example, an event that repeats at 7PM every day suddenly occurs at 6PM the day after daylight savings ends, even though it was stored in `dtstart` as UTC (with `Z` suffix).

I suppose a broader issue is whether it makes sense to use JS `Date` at all, given its limitations, or whether to provide the option to "hot swap" a library like Luxon (the successor to Moment, with robust timezone support) as @aramk 's (3 year old) fork does. Or if taking a 3rd party dependency is not an option, attempting to implement a proper "aware" `datetime` abstraction like the original Python RRULE is built on.

An even broader one is why did we choose to build a calendar in the first place, but it's too late for that, here we are. I can fully understand why devoting one's time and resources to this library long-term may not make the most sense without funding to support, but I'm willing to take a crack at it as long as I can get some pointers on what some acceptance criteria might be.

**Comment by aramk** [comment]
For what it's worth, I'm still using my forked version with the `TimeZoneDate` class in production since I created the PR. From what I can tell this library doesn't merge PRs often. https://github.com/jakubroztocil/rrule/pull/97

**Comment by davidgoli** [comment]
A simpler approach that might serve a common range of needs without needing robust, universal timezone support would be to provide an option to support either UTC or local. That would probably be enough for 90% of use cases requiring timezone support.

**Comment by davidgoli** [comment]
@aramk have you tried merging upstream into your fork lately? I took a look at it today and it seems complicated...

**Comment by aramk** [comment]
Yeah, I haven't bothered to try. I'm guessing it might be easier to re-apply my fixes on top with a new branch, but I'm not sure what fixes I'd be using from the latest version to warrant that.

**Comment by davidgoli** [comment]
Published in 2.4.0.
